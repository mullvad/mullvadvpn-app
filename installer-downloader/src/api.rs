//! Fetch information about app versions from the Mullvad API

/// See [module-level](self) docs.
#[async_trait::async_trait]
pub trait VersionInfoProvider {
    /// Return info about the stable version
    async fn get_version_info() -> anyhow::Result<VersionInfo>;
}

/// Contains information about all versions
#[derive(Debug, Clone)]
pub struct VersionInfo {
    /// Stable version info
    pub stable: Version,
    /// Beta version info
    pub beta: Option<Version>,
}

/// Contains information about a version for the current target
#[derive(Debug, Clone)]
pub struct Version {
    /// Version
    pub version: String,
    /// URLs to use for downloading the app installer
    pub urls: Vec<String>,
    /// Size of installer, in bytes
    pub size: usize,
    /// URLs pointing to app PGP signatures
    pub signature_urls: Vec<String>,
}

/// Use hardcoded URL to fetch installer
/// TODO: This is temporary
pub struct LatestVersionInfoProvider;

#[async_trait::async_trait]
impl VersionInfoProvider for LatestVersionInfoProvider {
    async fn get_version_info() -> anyhow::Result<VersionInfo> {
        Ok(VersionInfo {
            stable: Version {
                version: "2025.3".to_string(),
                urls: vec!["https://mullvad.net/en/download/app/exe/latest".to_owned()],
                size: 200 * 1024 * 1024,
                signature_urls: vec![
                    "https://mullvad.net/en/download/app/exe/latest/signature".to_owned()
                ],
            },
            beta: Some(Version {
                version: "2025.3-beta1".to_string(),
                urls: vec!["https://mullvad.net/en/download/app/exe/latest-beta".to_owned()],
                size: 200 * 1024 * 1024,
                signature_urls: vec![
                    "https://mullvad.net/en/download/app/exe/latest-beta/signature".to_owned(),
                ],
            }),
        })
    }
}
