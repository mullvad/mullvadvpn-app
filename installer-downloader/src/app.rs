//! This module implements the flow of downloading and verifying the app signature.

use std::path::PathBuf;

use crate::{
    fetch::{self, ProgressUpdater},
    verify::{AppVerifier, PgpVerifier},
};

#[derive(Debug)]
pub enum DownloadError {
    FetchSignature(anyhow::Error),
    FetchApp(anyhow::Error),
    Verification(anyhow::Error),
}

/// See the [module-level documentation](crate).
#[async_trait::async_trait]
pub trait AppDownloader {
    /// Download the app signature.
    async fn download_signature(&mut self) -> Result<(), DownloadError>;

    /// Download the app binary.
    async fn download_executable(&mut self) -> Result<(), DownloadError>;

    /// Verify the app signature.
    async fn verify(&mut self) -> Result<(), DownloadError>;
}

/// Download the app and signature, and verify the app's signature
pub async fn install_and_upgrade(mut downloader: impl AppDownloader) -> Result<(), DownloadError> {
    downloader.download_signature().await?;
    downloader.download_executable().await?;
    downloader.verify().await
}

#[derive(Clone)]
pub struct HttpAppDownloader<SigProgress, AppProgress> {
    signature_url: String,
    app_url: String,
    app_size: usize,
    signature_progress_updater: SigProgress,
    app_progress_updater: AppProgress,
    // TODO: set permissions
    tmp_dir: PathBuf,
}

impl<SigProgress, AppProgress> HttpAppDownloader<SigProgress, AppProgress> {
    const MAX_SIGNATURE_SIZE: usize = 1 * 1024;

    pub fn new(
        signature_url: &str,
        app_url: &str,
        app_size: usize,
        signature_progress_updater: SigProgress,
        app_progress_updater: AppProgress,
    ) -> Self {
        let tmp_dir = std::env::temp_dir();
        Self {
            signature_url: signature_url.to_owned(),
            app_url: app_url.to_owned(),
            app_size,
            signature_progress_updater,
            app_progress_updater,
            tmp_dir,
        }
    }
}

#[async_trait::async_trait]
impl<SigProgress: ProgressUpdater, AppProgress: ProgressUpdater> AppDownloader
    for HttpAppDownloader<SigProgress, AppProgress>
{
    async fn download_signature(&mut self) -> Result<(), DownloadError> {
        fetch::get_to_file(
            self.sig_path(),
            &self.signature_url,
            &mut self.signature_progress_updater,
            fetch::SizeHint::Maximum(Self::MAX_SIGNATURE_SIZE),
        )
        .await
        .map_err(DownloadError::FetchSignature)
    }

    async fn download_executable(&mut self) -> Result<(), DownloadError> {
        fetch::get_to_file(
            self.bin_path(),
            &self.app_url,
            &mut self.app_progress_updater,
            fetch::SizeHint::Exact(self.app_size),
        )
        .await
        .map_err(DownloadError::FetchApp)
    }

    async fn verify(&mut self) -> Result<(), DownloadError> {
        let bin_path = self.bin_path();
        let sig_path = self.sig_path();
        tokio::task::spawn_blocking(move || {
            PgpVerifier::verify(bin_path, sig_path).map_err(DownloadError::Verification)
        })
        .await
        .expect("verifier panicked")
    }
}

impl<SigProgress, AppProgress> HttpAppDownloader<SigProgress, AppProgress> {
    fn bin_path(&self) -> PathBuf {
        self.tmp_dir.join("temp.exe")
    }

    fn sig_path(&self) -> PathBuf {
        self.tmp_dir.join("temp.exe.sig")
    }
}
