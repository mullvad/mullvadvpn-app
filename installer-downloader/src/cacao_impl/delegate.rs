use std::sync::{Arc, Mutex};

use cacao::{control::Control, layout::Layout};

use super::ui::{Action, AppWindow};
use crate::delegate::{AppDelegate, AppDelegateQueue};

impl AppDelegate for AppWindow {
    type Queue = Queue;

    fn on_download<F>(&mut self, callback: F)
    where
        F: Fn() + Send + 'static,
    {
        let cb = Self::sync_callback(callback);
        self.download_button.button.set_action(move || {
            let cb = Action::DownloadClick(cb.clone());
            cacao::appkit::App::<super::ui::AppImpl, _>::dispatch_main(cb);
        });
    }

    fn on_cancel<F>(&mut self, callback: F)
    where
        F: Fn() + Send + 'static,
    {
        let cb = Self::sync_callback(callback);
        self.cancel_button.button.set_action(move || {
            let cb = Action::CancelClick(cb.clone());
            cacao::appkit::App::<super::ui::AppImpl, _>::dispatch_main(cb);
        });
    }

    fn set_status_text(&mut self, text: &str) {
        self.status_text.set_text(text);
    }

    fn set_download_text(&mut self, text: &str) {
        self.download_text.set_text(text);
    }

    fn show_download_progress(&mut self) {
        self.progress.set_hidden(false);
    }

    fn hide_download_progress(&mut self) {
        self.progress.set_hidden(true);
    }

    fn set_download_progress(&mut self, complete: u32) {
        self.progress.set_value(complete as f64);
    }

    fn show_download_button(&mut self) {
        self.download_button.button.set_hidden(false);
    }

    fn hide_download_button(&mut self) {
        self.download_button.button.set_hidden(true);
    }

    fn enable_download_button(&mut self) {
        self.download_button.button.set_enabled(true);
    }

    fn disable_download_button(&mut self) {
        self.download_button.button.set_enabled(false);
    }

    fn show_cancel_button(&mut self) {
        self.cancel_button.button.set_hidden(false);
    }

    fn hide_cancel_button(&mut self) {
        self.cancel_button.button.set_hidden(true);
    }

    fn enable_cancel_button(&mut self) {
        self.cancel_button.button.set_enabled(true);
    }

    fn disable_cancel_button(&mut self) {
        self.cancel_button.button.set_enabled(false);
    }

    fn show_beta_text(&mut self) {
        self.beta_link.set_hidden(false);
        self.beta_link_preface.set_hidden(false);
    }

    fn hide_beta_text(&mut self) {
        self.beta_link.set_hidden(true);
        self.beta_link_preface.set_hidden(true);
    }

    fn queue(&self) -> Self::Queue {
        Queue {}
    }
}

impl AppWindow {
    // NOTE: We need this horrible lock because Dispatcher demands Sync, but AppDelegate does not require Sync
    fn sync_callback(
        callback: impl Fn() + Send + 'static,
    ) -> Arc<Mutex<Box<dyn Fn() + Send + 'static>>> {
        Arc::new(Mutex::new(Box::new(move || callback())))
    }
}

/// This simply mutates the UI on the main thread using the GCD
pub struct Queue {}

impl AppDelegateQueue<AppWindow> for Queue {
    fn queue_main<F: FnOnce(&mut AppWindow) + 'static + Send>(&self, callback: F) {
        // NOTE: We need this horrible lock because Dispatcher demands Sync
        let cb: Mutex<Option<Box<dyn FnOnce(&mut AppWindow) + Send>>> =
            Mutex::new(Some(Box::new(callback)));
        cacao::appkit::App::<super::ui::AppImpl, _>::dispatch_main(Action::QueueMain(cb));
    }
}
