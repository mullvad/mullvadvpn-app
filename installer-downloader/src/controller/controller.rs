//! This module implements the actual logic performed by different UI components.

use super::{AppDelegate, AppDelegateQueue};

use crate::{
    api::{self, LatestVersionInfoProvider, VersionInfoProvider},
    app::{self, AppDownloader, AppDownloaderFactory, AppDownloaderParameters, HttpAppDownloader},
    fetch::ProgressUpdater,
};

use std::future::Future;
use tokio::sync::{mpsc, oneshot};

use super::ui_downloader::{UiAppDownloader, UiProgressUpdater};

/// This trait glues together different components needed to construct an app controller.
/// In particular, this lets us mock the downloader and version provider independently of the
/// UI implementation (some [AppDelegate]).
pub trait AppControllerProvider {
    type Delegate: AppDelegate + 'static;
    type DownloaderFactory: AppDownloaderFactory;
    type VersionInfoProvider: VersionInfoProvider;
}

/// Default implementation of [AppControllerProvider], using an actual HTTP client, and API version
/// fetcher.
pub struct DefaultAppControllerProvider<T> {
    _phantom: std::marker::PhantomData<T>,
}

impl<T> DefaultAppControllerProvider<T> {
    pub fn new() -> Self {
        Self {
            _phantom: std::marker::PhantomData,
        }
    }
}

impl<T: AppDelegate + 'static> AppControllerProvider for DefaultAppControllerProvider<T> {
    type Delegate = T;
    type DownloaderFactory = HttpAppDownloader<UiProgressUpdater<T>, UiProgressUpdater<T>>;
    type VersionInfoProvider = LatestVersionInfoProvider;
}

/// Actions handled by an async worker task in [handle_action_messages].
enum TaskMessage {
    SetVersionInfo(api::VersionInfo),
    BeginDownload,
    Cancel,
}

/// See the [module-level docs](self).
pub struct AppController {}

impl AppController {
    /// Initialize the app controller.
    pub fn initialize<T: AppControllerProvider + 'static>(delegate: &mut T::Delegate) {
        delegate.hide_download_button();
        delegate.hide_cancel_button();

        let (task_tx, task_rx) = mpsc::channel(1);
        tokio::spawn(handle_action_messages::<T>(delegate.queue(), task_rx));
        tokio::spawn(fetch_app_version_info::<T>(delegate, task_tx.clone()));
        Self::register_user_action_callbacks(delegate, task_tx);
    }

    fn register_user_action_callbacks<T: AppDelegate + 'static>(
        delegate: &mut T,
        task_tx: mpsc::Sender<TaskMessage>,
    ) {
        let tx = task_tx.clone();
        delegate.on_download(move |_delegate| {
            let _ = tx.try_send(TaskMessage::BeginDownload);
        });
        let tx = task_tx.clone();
        delegate.on_cancel(move |_delegate| {
            let _ = tx.try_send(TaskMessage::Cancel);
        });
    }
}

/// Background task that fetches app version data.
fn fetch_app_version_info<T: AppControllerProvider>(
    delegate: &mut T::Delegate,
    download_tx: mpsc::Sender<TaskMessage>,
) -> impl Future<Output = ()> {
    delegate.set_status_text("Fetching app version...");
    let queue = delegate.queue();

    async move {
        // TODO: handle errors, retry
        let Ok(version_info) = T::VersionInfoProvider::get_version_info().await else {
            queue.queue_main(move |self_| {
                self_.set_status_text("Failed to fetch version info");
            });
            return;
        };
        let _ = download_tx.try_send(TaskMessage::SetVersionInfo(version_info));
    }
}

/// Async worker that handles actions such as initiating a download, cancelling it, and updating
/// labels.
async fn handle_action_messages<T: AppControllerProvider>(
    queue: <T::Delegate as AppDelegate>::Queue,
    mut rx: mpsc::Receiver<TaskMessage>,
) {
    let mut version_info = None;
    let mut active_download = None;

    while let Some(msg) = rx.recv().await {
        match msg {
            TaskMessage::SetVersionInfo(new_version_info) => {
                let version_label = format!("Latest version: {}", new_version_info.stable.version);
                queue.queue_main(move |self_| {
                    self_.set_status_text(&version_label);
                    self_.show_download_button();
                });
                version_info = Some(new_version_info);
            }
            TaskMessage::BeginDownload => {
                if active_download.is_some() {
                    continue;
                }
                let Some(version_info) = version_info.clone() else {
                    continue;
                };

                let (tx, rx) = oneshot::channel();
                queue.queue_main(move |self_| {
                    // TODO: Select appropriate URLs
                    let Some(app_url) = version_info.stable.urls.first() else {
                        return;
                    };
                    let Some(signature_url) = version_info.stable.signature_urls.first() else {
                        return;
                    };
                    let app_size = version_info.stable.size;

                    self_.set_status_text("");
                    self_.disable_download_button();
                    self_.show_cancel_button();
                    self_.show_download_progress();

                    let new_delegated_downloader = |sig_progress, app_progress| {
                        HttpAppDownloader::new(AppDownloaderParameters {
                            signature_url: signature_url.to_owned(),
                            app_url: app_url.to_owned(),
                            app_size,
                            sig_progress,
                            app_progress,
                        })
                    };

                    let downloader = UiAppDownloader::new(self_, new_delegated_downloader);
                    let _ = tx.send(tokio::spawn(async move {
                        let _ = app::install_and_upgrade(downloader).await;
                    }));
                });
                active_download = rx.await.ok();
            }
            TaskMessage::Cancel => {
                let Some(active_download) = active_download.take() else {
                    continue;
                };
                active_download.abort();
                let _ = active_download.await;

                let version_label = if let Some(version_info) = &version_info {
                    format!("Latest version: {}", version_info.stable.version)
                } else {
                    "".to_owned()
                };

                queue.queue_main(move |self_| {
                    self_.set_status_text(&version_label);
                    self_.enable_download_button();
                    self_.hide_cancel_button();
                    self_.hide_download_progress();
                    self_.set_download_progress(0);
                });
            }
        }
    }
}
