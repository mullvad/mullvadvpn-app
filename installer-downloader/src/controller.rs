//! This module implements the actual logic performed by different UI components.

use crate::delegate::{AppDelegate, AppDelegateQueue};
use crate::resource;
use crate::ui_downloader::{UiAppDownloader, UiProgressUpdater};

use mullvad_update::api::Version;
use mullvad_update::{
    api::{self, VersionInfoProvider},
    app::{self, AppDownloaderFactory, AppDownloaderParameters},
};

use std::future::Future;
use tokio::sync::{mpsc, oneshot};

/// Actions handled by an async worker task in [handle_action_messages].
enum TaskMessage {
    SetVersionInfo(api::VersionInfo),
    BeginDownload,
    Cancel,
}

/// See the [module-level docs](self).
pub struct AppController {}

/// Public entry function for registering a [AppDelegate].
pub fn initialize_controller<T: AppDelegate + 'static>(delegate: &mut T) {
    use mullvad_update::api::LatestVersionInfoProvider;
    use mullvad_update::app::HttpAppDownloader;

    // App downloader (factory) to use
    type DownloaderFactory<T> = HttpAppDownloader<UiProgressUpdater<T>, UiProgressUpdater<T>>;
    // Version info provider to use
    type VersionInfoProvider = LatestVersionInfoProvider;

    AppController::initialize::<_, DownloaderFactory<T>, VersionInfoProvider>(delegate)
}

impl AppController {
    /// Initialize [AppController] using the provided delegate.
    ///
    /// Providing the downloader and version info fetcher as type arguments, they're decoupled from
    /// the logic of [AppController], allowing them to be mocked.
    pub fn initialize<Delegate, DownloaderFactory, VersionProvider>(delegate: &mut Delegate)
    where
        Delegate: AppDelegate + 'static,
        VersionProvider: VersionInfoProvider + 'static,
        DownloaderFactory: AppDownloaderFactory<
                SigProgress = UiProgressUpdater<Delegate>,
                AppProgress = UiProgressUpdater<Delegate>,
            > + 'static,
    {
        delegate.hide_download_progress();
        delegate.hide_download_button();
        delegate.hide_cancel_button();

        let (task_tx, task_rx) = mpsc::channel(1);
        tokio::spawn(handle_action_messages::<Delegate, DownloaderFactory>(
            delegate.queue(),
            task_rx,
        ));
        delegate.set_status_text(resource::FETCH_VERSION_DESC);
        tokio::spawn(fetch_app_version_info::<Delegate, VersionProvider>(
            delegate,
            task_tx.clone(),
        ));
        Self::register_user_action_callbacks(delegate, task_tx);
    }

    fn register_user_action_callbacks<T: AppDelegate + 'static>(
        delegate: &mut T,
        task_tx: mpsc::Sender<TaskMessage>,
    ) {
        let tx = task_tx.clone();
        delegate.on_download(move || {
            let _ = tx.try_send(TaskMessage::BeginDownload);
        });
        let tx = task_tx.clone();
        delegate.on_cancel(move || {
            let _ = tx.try_send(TaskMessage::Cancel);
        });
    }
}

/// Background task that fetches app version data.
fn fetch_app_version_info<Delegate, VersionProvider>(
    delegate: &mut Delegate,
    download_tx: mpsc::Sender<TaskMessage>,
) -> impl Future<Output = ()>
where
    Delegate: AppDelegate,
    VersionProvider: VersionInfoProvider,
{
    let queue = delegate.queue();

    async move {
        // TODO: handle errors, retry
        let Ok(version_info) = VersionProvider::get_version_info().await else {
            queue.queue_main(move |self_| {
                self_.set_status_text("Failed to fetch version info");
            });
            return;
        };
        let _ = download_tx.try_send(TaskMessage::SetVersionInfo(version_info));
    }
}

/// Async worker that handles actions such as initiating a download, cancelling it, and updating
/// labels.
async fn handle_action_messages<Delegate, DownloaderFactory>(
    queue: Delegate::Queue,
    mut rx: mpsc::Receiver<TaskMessage>,
) where
    Delegate: AppDelegate + 'static,
    DownloaderFactory: AppDownloaderFactory<
        SigProgress = UiProgressUpdater<Delegate>,
        AppProgress = UiProgressUpdater<Delegate>,
    >,
{
    let mut version_info = None;
    let mut active_download = None;

    while let Some(msg) = rx.recv().await {
        match msg {
            TaskMessage::SetVersionInfo(new_version_info) => {
                let version_label = format_latest_version(&new_version_info.stable);
                queue.queue_main(move |self_| {
                    self_.set_status_text(&version_label);
                    self_.show_download_button();
                });
                version_info = Some(new_version_info);
            }
            TaskMessage::BeginDownload => {
                if active_download.is_some() {
                    continue;
                }
                let Some(version_info) = version_info.clone() else {
                    continue;
                };

                let (tx, rx) = oneshot::channel();
                queue.queue_main(move |self_| {
                    // TODO: Select appropriate URLs
                    let Some(app_url) = version_info.stable.urls.first() else {
                        return;
                    };
                    let Some(signature_url) = version_info.stable.signature_urls.first() else {
                        return;
                    };
                    let app_size = version_info.stable.size;

                    self_.set_status_text("");
                    self_.hide_download_button();
                    self_.show_cancel_button();
                    self_.show_download_progress();

                    let downloader = DownloaderFactory::new_downloader(AppDownloaderParameters {
                        signature_url: signature_url.to_owned(),
                        app_url: app_url.to_owned(),
                        app_size,
                        sig_progress: UiProgressUpdater::new(self_.queue()),
                        app_progress: UiProgressUpdater::new(self_.queue()),
                    });

                    let ui_downloader = UiAppDownloader::new(self_, downloader);
                    let _ = tx.send(tokio::spawn(async move {
                        let _ = app::install_and_upgrade(ui_downloader).await;
                    }));
                });
                active_download = rx.await.ok();
            }
            TaskMessage::Cancel => {
                let Some(active_download) = active_download.take() else {
                    continue;
                };
                active_download.abort();
                let _ = active_download.await;

                let version_label = if let Some(version_info) = &version_info {
                    format_latest_version(&version_info.stable)
                } else {
                    "".to_owned()
                };

                queue.queue_main(move |self_| {
                    self_.set_status_text(&version_label);
                    self_.show_download_button();
                    self_.hide_cancel_button();
                    self_.hide_download_progress();
                    self_.set_download_progress(0);
                });
            }
        }
    }
}

fn format_latest_version(version: &Version) -> String {
    format!("{}: {}", resource::LATEST_VERSION_PREFIX, version.version)
}
