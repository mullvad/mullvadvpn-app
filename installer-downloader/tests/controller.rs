//! Tests for integrations between UI controller and other components
//!
//! The tests rely on `insta` for snapshot testing. If they fail due to snapshot assertions,
//! then most likely the snapshots need to be updated. The most convenient way to review
//! changes to, and update, snapshots are by running `cargo insta review`.

use insta::assert_yaml_snapshot;
use installer_downloader::controller::AppController;
use installer_downloader::delegate::{AppDelegate, AppDelegateQueue};
use installer_downloader::ui_downloader::UiAppDownloaderParameters;
use mullvad_update::api::{Version, VersionInfo, VersionInfoProvider};
use mullvad_update::app::{AppDownloader, DownloadError};
use mullvad_update::fetch::ProgressUpdater;
use std::sync::{Arc, LazyLock, Mutex};
use std::time::Duration;
use std::vec::Vec;

pub struct FakeVersionInfoProvider {}

static FAKE_VERSION: LazyLock<VersionInfo> = LazyLock::new(|| VersionInfo {
    stable: Version {
        version: "2025.1".to_owned(),
        urls: vec!["https://mullvad.net/fakeapp".to_owned()],
        size: 1234,
        signature_urls: vec!["https://mullvad.net/fakesig".to_owned()],
    },
    beta: None,
});

#[async_trait::async_trait]
impl VersionInfoProvider for FakeVersionInfoProvider {
    async fn get_version_info() -> anyhow::Result<VersionInfo> {
        Ok(FAKE_VERSION.clone())
    }
}

/// Downloader for which all steps immediately succeed
pub type FakeAppDownloaderHappyPath = FakeAppDownloader<true, true, true>;

/// Downloader for which the download step fails
pub type FakeAppDownloaderDownloadFail = FakeAppDownloader<true, false, false>;

/// Downloader for which all but the final verification step succeed
pub type FakeAppDownloaderVerifyFail = FakeAppDownloader<true, true, false>;

impl<const A: bool, const B: bool, const C: bool> From<UiAppDownloaderParameters<FakeAppDelegate>>
    for FakeAppDownloader<A, B, C>
{
    fn from(params: UiAppDownloaderParameters<FakeAppDelegate>) -> Self {
        FakeAppDownloader { params }
    }
}

/// Fake app downloader
///
/// Parameters:
/// * SIG_SUCCEED - whether fetching the signature succeeds
/// * EXE_SUCCEED - whether fetching the binary succeeds
/// * VERIFY_SUCCEED - whether verifying the signature succeeds
pub struct FakeAppDownloader<
    const SIG_SUCCEED: bool,
    const EXE_SUCCEED: bool,
    const VERIFY_SUCCEED: bool,
> {
    params: UiAppDownloaderParameters<FakeAppDelegate>,
}

#[async_trait::async_trait]
impl<const SIG_SUCCEED: bool, const EXE_SUCCEED: bool, const VERIFY_SUCCEED: bool> AppDownloader
    for FakeAppDownloader<SIG_SUCCEED, EXE_SUCCEED, VERIFY_SUCCEED>
{
    async fn download_signature(&mut self) -> Result<(), DownloadError> {
        self.params.sig_progress.set_url(&self.params.signature_url);
        self.params.sig_progress.set_progress(0.);
        if SIG_SUCCEED {
            self.params.sig_progress.set_progress(1.);
            Ok(())
        } else {
            Err(DownloadError::FetchSignature(anyhow::anyhow!(
                "fetching signature failed"
            )))
        }
    }

    async fn download_executable(&mut self) -> Result<(), DownloadError> {
        self.params.app_progress.set_url(&self.params.app_url);
        self.params.app_progress.set_progress(0.);
        if EXE_SUCCEED {
            self.params.app_progress.set_progress(1.);
            Ok(())
        } else {
            Err(DownloadError::FetchApp(anyhow::anyhow!(
                "fetching app failed"
            )))
        }
    }

    async fn verify(&mut self) -> Result<(), DownloadError> {
        if VERIFY_SUCCEED {
            Ok(())
        } else {
            Err(DownloadError::Verification(anyhow::anyhow!(
                "verification failed"
            )))
        }
    }
}

/// A fake queue that stores callbacks so that tests can run them later.
#[derive(Clone, Default)]
pub struct FakeQueue {
    callbacks: Arc<Mutex<Vec<Box<dyn FnOnce(&mut FakeAppDelegate) + Send>>>>,
}

impl FakeQueue {
    /// Run all queued callbacks on the given delegate.
    fn run_callbacks(&self, delegate: &mut FakeAppDelegate) {
        let mut callbacks = self.callbacks.lock().unwrap();
        for cb in callbacks.drain(..) {
            cb(delegate);
        }
    }
}

impl AppDelegateQueue<FakeAppDelegate> for FakeQueue {
    fn queue_main<F: FnOnce(&mut FakeAppDelegate) + 'static + Send>(&self, callback: F) {
        self.callbacks.lock().unwrap().push(Box::new(callback));
    }
}

/// A fake [AppDelegate]
#[derive(Default)]
pub struct FakeAppDelegate {
    /// Callback registered by `on_download`
    pub download_callback: Option<Box<dyn Fn() + Send>>,
    /// Callback registered by `on_cancel`
    pub cancel_callback: Option<Box<dyn Fn() + Send>>,
    /// State of delegate
    pub state: DelegateState,
    /// Queue used to simulate the main thread
    pub queue: FakeQueue,
}

/// A complete state of the UI, including its call history
#[derive(Default, serde::Serialize)]
pub struct DelegateState {
    pub status_text: String,
    pub download_text: String,
    pub download_button_visible: bool,
    pub cancel_button_visible: bool,
    pub cancel_button_enabled: bool,
    pub download_button_enabled: bool,
    pub download_progress: u32,
    pub download_progress_visible: bool,
    pub beta_text_visible: bool,
    /// Record of method calls.
    pub call_log: Vec<String>,
}

impl AppDelegate for FakeAppDelegate {
    type Queue = FakeQueue;

    fn on_download<F>(&mut self, callback: F)
    where
        F: Fn() + Send + 'static,
    {
        self.state.call_log.push("on_download".into());
        self.download_callback = Some(Box::new(callback));
    }

    fn on_cancel<F>(&mut self, callback: F)
    where
        F: Fn() + Send + 'static,
    {
        self.state.call_log.push("on_cancel".into());
        self.cancel_callback = Some(Box::new(callback));
    }

    fn set_status_text(&mut self, text: &str) {
        self.state
            .call_log
            .push(format!("set_status_text: {}", text));
        self.state.status_text = text.to_owned();
    }

    fn set_download_text(&mut self, text: &str) {
        self.state
            .call_log
            .push(format!("set_download_text: {}", text));
        self.state.download_text = text.to_owned();
    }

    fn show_download_progress(&mut self) {
        self.state.call_log.push("show_download_progress".into());
        self.state.download_progress_visible = true;
    }

    fn hide_download_progress(&mut self) {
        self.state.call_log.push("hide_download_progress".into());
        self.state.download_progress_visible = false;
    }

    fn set_download_progress(&mut self, complete: u32) {
        self.state
            .call_log
            .push(format!("set_download_progress: {}", complete));
        self.state.download_progress = complete;
    }

    fn show_download_button(&mut self) {
        self.state.call_log.push("show_download_button".into());
        self.state.download_button_visible = true;
    }

    fn hide_download_button(&mut self) {
        self.state.call_log.push("hide_download_button".into());
        self.state.download_button_visible = false;
    }

    fn enable_download_button(&mut self) {
        self.state.call_log.push("enable_download_button".into());
        self.state.download_button_enabled = true;
    }

    fn disable_download_button(&mut self) {
        self.state.call_log.push("disable_download_button".into());
        self.state.download_button_enabled = false;
    }

    fn show_cancel_button(&mut self) {
        self.state.call_log.push("show_cancel_button".into());
        self.state.cancel_button_visible = true;
    }

    fn hide_cancel_button(&mut self) {
        self.state.call_log.push("hide_cancel_button".into());
        self.state.cancel_button_visible = false;
    }

    fn enable_cancel_button(&mut self) {
        self.state.call_log.push("enable_cancel_button".into());
        self.state.cancel_button_enabled = true;
    }

    fn disable_cancel_button(&mut self) {
        self.state.call_log.push("disable_cancel_button".into());
        self.state.cancel_button_enabled = false;
    }

    fn show_beta_text(&mut self) {
        self.state.call_log.push("show_beta_text".into());
        self.state.beta_text_visible = true;
    }

    fn hide_beta_text(&mut self) {
        self.state.call_log.push("hide_beta_text".into());
        self.state.beta_text_visible = false;
    }

    fn queue(&self) -> Self::Queue {
        self.queue.clone()
    }
}

/// Test that the flow starts by fetching app version data
#[tokio::test(start_paused = true)]
async fn test_fetch_version() {
    let mut delegate = FakeAppDelegate::default();
    AppController::initialize::<_, FakeAppDownloaderHappyPath, FakeVersionInfoProvider>(
        &mut delegate,
    );

    // The app should start out by fetching the current app version
    assert_yaml_snapshot!(delegate.state);

    tokio::time::sleep(Duration::from_secs(1)).await;

    // Run UI updates to display the fetched version
    let queue = delegate.queue.clone();
    queue.run_callbacks(&mut delegate);

    // The download button and current version should be displayed
    assert_yaml_snapshot!(delegate.state);
}

/// Test that the on_download callback gets registered and, when invoked,
/// properly updates the UI.
#[tokio::test(start_paused = true)]
async fn test_download() {
    let mut delegate = FakeAppDelegate::default();
    AppController::initialize::<_, FakeAppDownloaderHappyPath, FakeVersionInfoProvider>(
        &mut delegate,
    );

    // Wait for the version info
    tokio::time::sleep(Duration::from_secs(1)).await;

    let queue = delegate.queue.clone();
    queue.run_callbacks(&mut delegate);

    // The download button should be available
    assert_yaml_snapshot!(delegate.state);

    // Initiate download
    let cb = delegate
        .download_callback
        .take()
        .expect("no download callback registered");
    cb();

    tokio::time::sleep(Duration::from_secs(1)).await;

    // Run queued actions
    let queue = delegate.queue.clone();
    queue.run_callbacks(&mut delegate);

    // We should see download progress, and cancellation
    assert_yaml_snapshot!(delegate.state);

    // Wait for download
    tokio::time::sleep(Duration::from_secs(1)).await;

    let queue = delegate.queue.clone();
    queue.run_callbacks(&mut delegate);

    // Everything including verification should have succeeded
    assert_yaml_snapshot!(delegate.state);
}

/// Test that the install aborts if verification fails
#[tokio::test(start_paused = true)]
async fn test_failed_verification() {
    let mut delegate = FakeAppDelegate::default();
    AppController::initialize::<_, FakeAppDownloaderVerifyFail, FakeVersionInfoProvider>(
        &mut delegate,
    );

    // Wait for the version info
    tokio::time::sleep(Duration::from_secs(1)).await;

    let queue = delegate.queue.clone();
    queue.run_callbacks(&mut delegate);

    // Initiate download
    let cb = delegate
        .download_callback
        .take()
        .expect("no download callback registered");
    cb();

    tokio::time::sleep(Duration::from_secs(1)).await;

    // Wait for queued actions to complete
    let queue = delegate.queue.clone();
    queue.run_callbacks(&mut delegate);

    tokio::time::sleep(Duration::from_secs(1)).await;

    let queue = delegate.queue.clone();
    queue.run_callbacks(&mut delegate);

    // Verification failed
    assert_yaml_snapshot!(delegate.state);
}
