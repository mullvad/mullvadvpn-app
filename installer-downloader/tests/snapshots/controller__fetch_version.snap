---
source: installer-downloader/tests/controller.rs
expression: delegate.state
snapshot_kind: text
---
status_text: Loading version details...
download_text: ""
download_button_visible: true
cancel_button_visible: false
cancel_button_enabled: false
download_button_enabled: false
download_progress: 0
download_progress_visible: false
beta_text_visible: false
call_log:
  - hide_download_progress
  - show_download_button
  - disable_download_button
  - hide_cancel_button
  - hide_beta_text
  - "set_status_text: Loading version details..."
  - on_download
  - on_cancel
